site/docs/spec.md

```diff
@@ -208,136 +256,202 @@ Notes:
 
 For example, an `events` table with a timestamp column named `ts` that is partitioned by `ts_day=day(ts)` is queried by users with ranges over the timestamp column: `ts > X`. The inclusive projection is `ts_day >= day(X)`, which is used to select files that may have matching rows. Note that, in most cases, timestamps just before `X` will be included in the scan because the file contains rows that match the predicate and rows that do not match the predicate.
 
-Notes:
+Scan predicates are also used to filter data and delete files using column bounds and counts that are stored by field id in manifests. The same filter logic can be used for both data and delete files because both stored metrics of the rows either inserted or deleted. If metrics show that a delete file has no rows that match a scan predicate, it may be ignored just as a data file would be ignored [2].
 
-1. An alternative, *strict projection*, creates a partition predicate that will match a file if all of the rows in the file must match the scan predicate. These projections are used to calculate the residual predicates for each file in a scan.
+Data files that match the query filter must be read by the scan.
 
-#### Manifest Lists
+Delete files that match the query filter must be applied to data files at read time, limited by the scope of the delete file using the following rules.
 
-Snapshots are embedded in table metadata, but the list of manifests for a snapshot can be stored in a separate manifest list file.
+* A _position_ delete file must be applied to a data file when all of the following are true:
+    - The data file's sequence number is _less than or equal to_ the delete file's sequence number
+    - The data file's partition (both spec and partition values) is equal to the delete file's partition
+* An _equality_ delete file must be applied to a data file when all of the following are true:
+    - The data file's sequence number is _strictly less than_ the delete's sequence number
+    - The data file's partition (both spec and partition values) is equal to the delete file's partition _or_ the delete file's partition spec is unpartitioned
 
-A manifest list encodes extra fields that can be used to avoid scanning all of the manifests in a snapshot when planning a table scan. 
+In general, deletes are applied only to data files that are older and in the same partition, except for two special cases:
 
-Manifest list files store `manifest_file`, a struct with the following fields:
+* Equality delete files stored with an unpartitioned spec are applied as global deletes. Otherwise, delete files do not apply to files in other partitions.
+* Position delete files must be applied to data files from the same commit, when the data and delete file sequence numbers are equal. This allows deleting rows that were added in the same commit.
```

Comment:
Streaming writes need this. Imagine a Flink job that is replacing rows in batches as they come in by some key:

```
(id=5, data="a")
...
(id=5, data="b")
...
(id=5, data="c")
```

To encode the first, it would use an equality delete for `id=5`, but after that the sink can't delete the newly added row with an equality delete against the same batch because equality deletes would remove all copies, even the newly written one. (That's why equality deletes only apply to strictly older files.)

To delete a row that is in the same batch (same sequence number), we have to use a position delete because those can target a specific row in a file. The writer needs to keep track of possible conflicts and encode deletes against the data files it is writing, but at least it doesn't need to buffer for the entire batch, then remove duplicates, and write all of the rows at once.